pandas/core/internals/construction.py

```diff
@@ -574,12 +574,30 @@ def _convert_object_array(content, columns, coerce_float=False, dtype=None):
     if columns is None:
         columns = ibase.default_index(len(content))
     else:
-        if len(columns) != len(content):  # pragma: no cover
+        is_mi_list = isinstance(columns, list) and all(
```

Comment:
can you add some comments on what this section is about (list-of-lists)

Reply:
I would move this whole check to _validate_columns, and remove the columns argument from _convert_object_array (and do it as an independent check before).

Reply:
Nice idea! @jreback 
moving the check out of `_convert_object_array` makes the scope of `_convert_object_array` clearer, and dealing with `columns` and `content` separately seem nicer

but if only moving the whole check part out will not be able to remove `columns` argument from `_convert_object_array`, because it also creates column names using length of contetn if `columns` is `None`. So i made a method called `_validate_or_indexify_columns` method (name could be improved if made method clearer ^^), and will do two things: 1. if `columns` is `None`, then indexify it with `default_index`, 2. otherwise, go through the checks.

What do you think of this change, does it make code clearer? @jreback any feedbacks are very welcome!